### PR TITLE
Scenario select cleanup

### PR DESCRIPTION
**Pull request by duncanspumpkin**
Addresses outstanding issues from #19693. It was a bit fiddly so decided best to do myself.

### CODE REVIEW

File: src/openrct2-ui/windows/ScenarioSelect.cpp [hunk 1]
```diff
@@ -594,224 +497,310 @@ class ScenarioSelectWindow final : public Window
...
+    }
 
     int32_t screenWidth = ContextGetWidth();
     int32_t screenHeight = ContextGetHeight();
     ScreenCoordsXY screenPos = { (screenWidth - WW) / 2, std::max(TOP_TOOLBAR_HEIGHT + 1, (screenHeight - WH) / 2) };
     window = WindowCreate<ScenarioSelectWindow>(WindowClass::ScenarioSelect, screenPos, WW, WH, 0);
-    window->widgets = window_scenarioselect_widgets;
-    window->highlighted_scenario = nullptr;
-
+    window->Init(callback, titleEditor);
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/ScenarioSelect.cpp:804]: There are better ways of doing this, either you could implement a constructor, and then move it during `WindowCreate`:
https://github.com/OpenRCT2/OpenRCT2/blob/e1f2de988eb3f4b76bd19c233631d24db730e180/src/openrct2-ui/windows/SavePrompt.cpp#L241-L244
or with the variadic option (ctor arguments coming last):
https://github.com/OpenRCT2/OpenRCT2/blob/056af36cdfefbdb42f78dbbca96a6a20fe4423b6/src/openrct2-ui/windows/LoadSave.cpp#L1053-L1054

File: src/openrct2-ui/windows/ScenarioSelect.cpp [hunk 2]
```diff
@@ -594,224 +497,310 @@ class ScenarioSelectWindow final : public Window
...
-            }
-        }
-        else
-        {
-            if (w->selected_tab <= SCENARIO_CATEGORY_EXPERT)
+            const ScenarioIndexEntry* scenario = ScenarioRepositoryGetByIndex(i);
+
+            if (!IsScenarioVisible(*scenario))
+                continue;
+            if (_titleEditor && scenario->SourceGame == ScenarioSource::Other)
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/ScenarioSelect.cpp:559]: Use `IsTitleEditor()` instead

File: src/openrct2-ui/windows/ScenarioSelect.cpp [hunk 3]
```diff
@@ -594,224 +497,310 @@ class ScenarioSelectWindow final : public Window
...
+                continue;
+
+            // Category heading
+            StringId headingStringId = STR_NONE;
+            if (gConfigGeneral.ScenarioSelectMode == SCENARIO_SELECT_MODE_ORIGIN || _titleEditor)
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/ScenarioSelect.cpp:564]: Use `IsTitleEditor()` instead

File: src/openrct2-ui/windows/ScenarioSelect.cpp [hunk 4]
```diff
@@ -594,224 +497,310 @@ class ScenarioSelectWindow final : public Window
...
                 }
             }
+        }
+    }
 
-            // If scenario is Mega Park, keep a reference to it
-            if (scenario->ScenarioId == SC_MEGA_PARK)
+    bool IsScenarioVisible(const ScenarioIndexEntry& scenario) const
+    {
+        if (gConfigGeneral.ScenarioSelectMode == SCENARIO_SELECT_MODE_ORIGIN || _titleEditor)
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/ScenarioSelect.cpp:669]: Use `IsTitleEditor()` instead

File: src/openrct2-ui/windows/ScenarioSelect.cpp [hunk 5]
```diff
@@ -594,224 +497,310 @@ class ScenarioSelectWindow final : public Window
...
-        {
-            // Remove mega park
-            _listItems.pop_back();
+        if (gConfigGeneral.ScenarioSelectMode != SCENARIO_SELECT_MODE_ORIGIN)
+            return false;
+        if (!gConfigGeneral.ScenarioUnlockingEnabled)
+            return false;
+        if (selected_tab >= 6)
+            return false;
+        if (_titleEditor)
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/ScenarioSelect.cpp:699]: Use `IsTitleEditor()` instead

**Review by tupaschoal — COMMENTED** [review]
Change looks ok, though the first commit has a bonkers title. I have some suggestions to improve code readability, but that should be all

(Haven't tested though)

File: src/openrct2-ui/windows/ScenarioSelect.cpp (hunk 1 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/ScenarioSelect.cpp:804]: For all the other windows we have used an Init function for now at least. Perhaps on a second pass

**Comment by duncanspumpkin** [comment]
> Change looks ok, though the first commit has a bonkers title. I have some suggestions to improve code readability, but that should be all
> 
> (Haven't tested though)

I've tested it a bit but not the titleeditor parts.

**Review by tupaschoal — APPROVED** [review]

**Pull request merged**